test/plugins/cookieauth.js

```diff
@@ -0,0 +1,984 @@
+          .get(DBNAME)
+          .reply(200, {doc_count: 0});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        if (!process.env.NOCK_OFF) {
+          assert.equal(resp.request.headers.cookie, MOCK_COOKIE);
+        }
+        assert.equal(resp.statusCode, 200);
+        assert.ok(data.indexOf('"doc_count":0') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('performs request and returns 500 response', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER)
+          .get(DBNAME)
+          .reply(500, {error: 'internal_server_error', reason: 'Internal Server Error'});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        assert.equal(resp.request.headers.cookie, MOCK_COOKIE);
+        assert.equal(resp.statusCode, 500);
+        assert.ok(data.indexOf('"error":"internal_server_error"') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('performs request and returns error', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(200, {ok: true}, MOCK_SET_COOKIE_HEADER)
+          .get(DBNAME)
+          .replyWithError({code: 'ECONNRESET', message: 'socket hang up'});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err.code, 'ECONNRESET');
+        assert.equal(err.message, 'socket hang up');
+        mocks.done();
+        done();
+      });
+    });
+
+    it('returns client error 401 response on authentication failure', function(done) {
+      if (process.env.NOCK_OFF) {
+        this.skip();
+      }
+
+      var mocks = nock(SERVER)
+          .post('/_session', {name: ME, password: PASSWORD})
+          .reply(401, {error: 'unauthorized', reason: 'Unauthorized'});
+
+      var cloudantClient = new Client({ plugin: 'cookieauth' });
+      var req = { url: SERVER + DBNAME, method: 'GET' };
+      cloudantClient.request(req, function(err, resp, data) {
+        assert.equal(err, null);
+        assert.equal(resp.statusCode, 401);
+        assert.ok(data.indexOf('"error":"unauthorized"') > -1);
+        mocks.done();
+        done();
+      });
+    });
+
+    it('returns client error 403 response on authentication failure', function(done) {
```

Comment:
I think this test is invalid.
We shouldn't have any special handling for `403` when requesting session, because I don't think it is a valid response. I think this test is valid from the perspective of getting an error code other than `200` or `401` and dealing with it correctly, but it should probably be a `5XX` for that to avoid confusion with the `403` that might be received on non-`_session` responses.

Reply:
So, the flow for making a request is something like this?
1. Clear expired cookies for idx _user.cloudant.com_.
2. If `cookies.length == 0` then make a `POST /_session`, on response:
   - `200`: continue
   - error || `5xx`: retry
   - default: continue _(without cookie)_
3. Make client request.

Reply:
1. Clear expired cookies for idx _user.cloudant.com_.
1. If `cookies.length == 0` then make a `POST /_session`, on response:
    1. `200` - store cookie, proceed with request
    1. `401` - make no more `_session` requests, continue without cookie
    1. `4xx/5xx`/error - behaves as 401 in the other libs, but I could see an argument for retrying
1. Make client request; on response:
    1. `401` || `403 {reason: credentials_expired}` - retry client request (with new `_session` request).
    1. Default pass response back